test/integration/src/test/java/io/pravega/test/integration/endtoendtest/EndToEndStatsTest.java

```diff
@@ -109,32 +115,39 @@ public void testStatsCount() throws Exception {
         EventStreamWriter<String> test = clientFactory.createEventWriter("test", new JavaSerializer<>(),
                 EventWriterConfig.builder().transactionTimeoutTime(10000).build());
 
+        String[] tags = segmentTags(StreamSegmentNameUtils.getQualifiedStreamSegmentName("test", "test", 0L));
+
         for (int i = 0; i < 10; i++) {
             test.writeEvent("test").get();
         }
-        assertEventuallyEquals(10, () -> statsRecorder.getSegments().get(StreamSegmentNameUtils.getQualifiedStreamSegmentName("test", "test", 0L)).get(), 2000);
+        assertEventuallyEquals(10, () -> (int) (statsRecorder.getRegistry().counter(SEGMENT_WRITE_EVENTS, tags).count()), 2000);
```

Comment:
Why the timeouts (last argument) in all the invocations to `assertEventuallyEquals` are different? A 1 or 2 seconds timeout seems reasonable for all the cases, right?

Reply:
@RaulGracia I repeated the test 517 times and finally triggered the problem. The bottom line is that HashMap is not thread safe, it shouldn't be used in a multiple thread env - and indeed the problem happened.

Yes 2 second for regular write assertion, but 10 seconds for transaction write to play it safe (I didn't set these 2 values but I'm ok with the values set).
For the additional Write Byte assertion I added, however, I give it 100ms max waiting instead of 2 second, because this assertion is after Write Event assertion, which should have already waited sufficient time.

Reply:
`HashMap` is not thread-safe, you are right, in the same way that the recently added `HashSet` is neither thread-safe (and you are not using synchronization), right? In the changes of this PR, I see that now there is no "removal" of elements from the `metrics` data structure, which may alleviate the problem. But note that if you had left the "remove" operation (or other update concurrent operations on that data structure), we may probably observe problems too.

Reply:
The strong reference place holder (HashSet) has no consumer hence no concurrent concern of it, however it doesn't hurt to synchronize it; also I changed the HashSet to LinkedList to improve the performance and reduce the memory footprint a bit.
I also changed the name "metrics" to "references" to emphasis it is a place holder, instead of a metrics registry, so there is no need to manage its life-cycle. (The test case will create about 10+ metrics objects, and all the objects will be released when tear down the test case)